### PR TITLE
Cal py3 [Calibre-Web: move to Python 3] ...uncovers 2 Calibre-Web issues: slow shutdown & "Perform Update" button risks

### PR DESCRIPTION
**Pull request by jvonau**
### Fixes Bug
move to python3
### Description of changes proposed in this pull request.
move calibre-web to python3 #2231 
### Smoke-tested in operating system.
raspbian-10

### CODE REVIEW

File: roles/calibre-web/tasks/install.yml [hunk 1]
```diff
@@ -42,21 +42,21 @@
...
   when: internet_available | bool
 
 # 2020-02-02: IS PYTHON 3 POSSIBLE ABOVE AND BELOW?
 
-- name: Symlink {{ calibreweb_venv_path }}/vendor -> {{ calibreweb_venv_path }}/lib/python2.7/site-packages - to keep {{ calibreweb_venv_path }}/cps.py happy
+- name: Symlink {{ calibreweb_venv_path }}/vendor -> {{ calibreweb_venv_path }}/lib/python2.7/site-packages - to keep {{ calibreweb_venv_path }}/cps.py happy?
   file:
-    src: "{{ calibreweb_venv_path }}/lib/python2.7/site-packages"
-    path: "{{ calibreweb_venv_path }}/vendor"    # /usr/local/calibre-web
+    src: "{{ calibreweb_venv_path }}/lib/python3.7/site-packages"
```
Comment (holta) [inline roles/calibre-web/tasks/install.yml]: Ubuntu 18.04 uses Python 3.6.x  — whereas the above line mentions 3.7 — how should/would we accommodate OS's that do not offer 3.7.x ?

(Ubuntu 20.04 arriving soon may include Python 3.8.x, as another example.)

Comment (jvonau) [reply roles/calibre-web/tasks/install.yml]: get upstream to correct the layout as not to need vendor.

**Comment by holta** [comment]
@OzzieIsaacs can you take a quick look at this short PR towards advising Internet-in-a-Box on how we should use Calibre-Web with Python 3?

(Rather than Python 2.7, as our virtual environment uses today!)

Thanks !

**Comment by OzzieIsaacs** [comment]
For python2 and python3 compatible software the recommended shebang line is to my knowledge a simple "python" as calibre-web uses it. On Raspian 10 the python command points to python 2 (like a simple `python --version` shows). So if you installed all dependencies with python3, a `python cps.py`, or a simple `./cps.py` command will fail (because tornado, the used wsgi server is no longer compatible with python2, like in my issue janeczku/calibre-web#993). From my point of view a systemd script which is explicitly starting python3 -> `python3 cps.py` should do it. 

> get upstream to correct the layout as not to need vendor.

You don't need to install everything to the "vendor" folder, it's just an easy way around venv, different dependencies versions and so on in the system. It's basically a leftover from the beginning of the project. I think it was made, so you could not destroy anything in your system, if something goes wrong, you just delete the whole calibre-web path and everything is as before. So a simple `pip3 install requirements.txt` will do it, or use your virtual environment.

I hope I answered the right question, about the ansible (?) file changes, I don't know.

**Comment by holta** [comment]
@jvonau says run `systemctl stop calibre-web` to avoid grief before testing this.

**Comment by jvonau** [comment]
can test on preexisting install with
`sudo systemctl stop calibre-web`
`./runrole --reinstall calibre-web`

File: roles/calibre-web/tasks/install.yml [hunk 2]
```diff
@@ -42,23 +42,15 @@
 #  ignore_errors: True
 ##
 # Implementing this with Ansible command module for now.
-- name: Download Calibre-Web dependencies (using pip) into python2.7 virtual environment {{ calibreweb_venv_path }}
-  pip:
+- name: Download Calibre-Web dependencies (using pip) into python3.7 virtual environment {{ calibreweb_venv_path }}
```
Comment (holta) [inline roles/calibre-web/tasks/install.yml]:
```suggestion
- name: Download Calibre-Web dependencies (using pip) into python3 virtual environment {{ calibreweb_venv_path }}
```

**Comment by holta** [comment]
> can test on preexisting install with
> `sudo systemctl stop calibre-web`
> `./runrole --reinstall calibre-web`

When trying the above on Ubuntu 19.10, the following error results:

```
TASK [calibre-web : Install Calibre-Web if 'calibreweb_installed' not defined, e.g. in /etc/iiab/iiab_state.yml] ***
fatal: [127.0.0.1]: FAILED! => {"reason": "couldn't resolve module/action 'pip3'. This often indicates a misspelling, missing collection, or incorrect module path.\n\nThe error appears to be in '/opt/iiab/iiab/roles/calibre-web/tasks/install.yml': line 45, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n# Implementing this with Ansible command module for now.\n- name: Download Calibre-Web dependencies (using pip) into python3.7 virtual environment {{ calibreweb_venv_path }}\n  ^ here\nWe could be wrong, but this one looks like it might be an issue with\nmissing quotes. Always quote template expression brackets when they\nstart a value. For instance:\n\n    with_items:\n      - {{ foo }}\n\nShould be written as:\n\n    with_items:\n      - \"{{ foo }}\"\n"}
```

Which I'm presuming is b/c Ansible does not have a pip3 module.

Ansible's [pip module documentation](https://docs.ansible.com/ansible/latest/modules/pip_module.html) discusses parameter `executable: pip3.3` &mdash; is that useful...or no?

**Comment by holta** [comment]
> can test on preexisting install with
> `sudo systemctl stop calibre-web`
> `./runrole --reinstall calibre-web`

Above test now works, over a pre-existing BIG-sized IIAB install (on Ubuntu 19.10).

@jvonau: should the equivalent of `systemctl stop calibre-web` happen at/near the top of [calibre-web/tasks/install.yml](https://github.com/jvonau/iiab/blob/cal_py3/roles/calibre-web/tasks/install.yml) &mdash; to "avoid grief" as you hinted earlier in the day?

**Comment by jvonau** [comment]
On related note, how long is it taking to stop calibre-web service?  I noticed that the service is reaching the systemd stock 90 second timeout when stopping the service then the process is forcefully killed. I believe this situation exists with python2 also, noticed the long delay shutting down for the upgrade but didn't dawn on me that the service was being abruptly shutdown.
>pi@box:/opt/iiab/iiab $ sudo systemctl status calibre-web.service --no-pager -l
● calibre-web.service
     Loaded: loaded (]8;;file://box.lan/etc/systemd/system/calibre-web.service/etc/systemd/system/calibre-web.service]8;;; enabled; vendor preset: enabled)
     Active: failed (Result: timeout) since Fri 2020-02-14 07:12:47 GMT; 53s ago
    Process: 22518 ExecStart=/usr/local/calibre-web-py3/bin/python3 /usr/local/calibre-web-py3/cps.py -p /library/calibre-web/config/app.db (code=killed, signal=KILL)
   Main PID: 22518 (code=killed, signal=KILL)
Feb 13 18:10:33 box.lan systemd[1]: Started calibre-web.service.
Feb 14 07:11:17 box.lan systemd[1]: Stopping calibre-web.service...
Feb 14 07:12:47 box.lan systemd[1]: calibre-web.service: State 'stop-sigterm' timed out. Killing.
Feb 14 07:12:47 box.lan systemd[1]: calibre-web.service: Killing process 22518 (python3) with signal SIGKILL.
Feb 14 07:12:47 box.lan systemd[1]: calibre-web.service: Main process exited, code=killed, status=9/KILL
Feb 14 07:12:47 box.lan systemd[1]: calibre-web.service: Failed with result 'timeout'.
Feb 14 07:12:47 box.lan systemd[1]: Stopped calibre-web.service.

The concern would be the db backend may not be finished before being killed by systemd as being misbehaving. Think it might be worthwhile to extend the shutdown timeout to see if that improves the situation.

**Comment by jvonau** [comment]
Doubling the timeout to 180 isn't enough
>Feb 14 07:27:44 box.lan systemd[1]: Started Calibre-Web.
Feb 14 07:28:07 box.lan systemd[1]: Stopping Calibre-Web...
Feb 14 07:31:07 box.lan systemd[1]: calibre-web.service: State 'stop-sigterm' timed out. Killing.

**Comment by jvonau** [comment]
> @jvonau: should the equivalent of `systemctl stop calibre-web` happen at/near the top of calibre-web/tasks/install.yml — to "avoid grief" as you hinted earlier in the day?

No, I would be concerned that ansible might interpret the systemd timeout stopping the service as a failure. I would like broader feedback on the stopping time on current python2 installs is see if this is something new with python3 or something never noticed before. Could explain some longer delays when rebooting/shutdown.

**Comment by OzzieIsaacs** [comment]
Normally calibre-web should finish within a few (1 or 2 at most) seconds (at least when I'm debugging it). Calibre-web uses two threads, the main one for handing the web requests and interacting with the calibre-database and the second one (worker thread) for emailing/converting books and doing the selfupdate. The main thread normally finishs immediately, so there is no risk in coruppting the database. The worker thread normally takes the few seconds for finshing. I'll try to find out why this is a problem with systemd (I think I saw this also recently with my calibre-web instance on my server during shutdown)

**Comment by holta** [comment]
@OzzieIsaacs thank you for investigating if you have time!

I can say from my experience with teachers in (many!) very poor countries that this "shut down time" really DOES matter, given the burdens that low-paid teachers are forced to carry in almost every country.

Which is just to re-affirm that this is definitely worth understanding and if possible fixing, i.e. if there are further details you need from Internet-in-a-Box to isolate/diagnose let us know what you need?

**Comment by holta** [comment]
> I had time to take a look at it last weekend, and I found the problem and I could fix it.

Many Thanks if you can @OzzieIsaacs.

Teachers would greatly appreciate this, particularly in countries where there's a lack of electrical power (where they need to turn off their battery-powered systems frequently!)

But if this is not possible in February here, possibly you can consider this for Calibre-Web 0.6.6 ?

**Comment by OzzieIsaacs** [comment]
I knew you would ask me for another release. 
There is another big problem, which causes a lot of trouble, so I think I will do a release this weekend to solve this (incompatibility with new "werkzeug" version 1.0.0)

**Comment by holta** [comment]
> I will do a release this weekend

Many teachers in Mexico's poorest states (Oaxaca & Chiapas) will thank you next week and the week after if you succeed.  But if not, don't sweat it, and we'll do better next time!

@OzzieIsaacs two related side questions:

1) Is there a recommended list of things to do, to completely & hygienically uninstall Calibre-Web?

2) Likewise, is there a recommended way to upgrade Calibre-Web?

   FYI Calibre-Web's "UPDATE" button at the bottom of http://box/books/admin/view has not worked for me, in fact damaged the installation of Calibre-Web, forcing me to reinstall all of Calibre-Web!

   I can handle that without crying (no tears!) but looking forward...we really don't want teachers to click this button and damage their Calibre-Web installation...do you have any recommendations here...to hide this button at minimum...if it continues to risk being dangerous like this?

**Comment by OzzieIsaacs** [comment]
Question 1:
- Delete the root folder of calibre-web, everything is stored within this folder, unless you didn't change location of logfiles and settingsfile via command line. Depending on how you installed the dependencies you have to take care about them: delete the virtual environment folder, or the vendor folder (should be within calibre-web folder). If dependencies are installed systemwide via "pip" in the sites-packages folder you have to uninstall it (difficult to handle, as other programs also could need this). 
Don't forget to delete systemd scripts  and other files, you might created.
Question 2:
The updater is quite stupid, just takes all the files from the new version (included in the zip file downloaded from the github page), replaces all files in the folder by the new files and deletes all files/folders not included anymore in the new zip file. EXCEPT: The settings file, the log files, the settings files for the gdrive integrtion and the vendor folder. I will add up the "venv" folder (I thought I already did this). I will try to improve the logic, so it will exclude folders within calibre.web where dependencies are loaded from. 
For more ideas I need more details what exactly went wrong beside "...has not worked for me, in fact damaged the installation of Calibre-Web..."

**Comment by holta** [comment]
> For more ideas I need more details what exactly went wrong beside "...has not worked for me, in fact damaged the installation of Calibre-Web..."

OK, a bit more below.  Tell me what else you need.  Just to be clear: Internet-in-a-Box installations are usually offline, so hiding the "Perform Update" button might be the best option?!  In any case, we just really don't want teachers/schools/clinics damaging their Calibre-Web installations by accident, when they are occasionally online, and very tempted to click the button (:

So...when using Calibre-Web's "Perform Update" button to try to upgrade from 0.6.4 to 0.6.5, this is the result for now...

1) http://box/books shows:

> Service Unavailable
The server is temporarily unable to service your request due to maintenance downtime or capacity problems. Please try again later.
>
> Apache/2.4.18 (Ubuntu) Server at 192.168.0.123 Port 80

2) systemd service 'calibre-web' cannot be started, as seen here:

> root@box:~# systemctl status calibre-web
● calibre-web.service
   Loaded: loaded (/etc/systemd/system/calibre-web.service; enabled; vendor preset: enabled)
   Active: failed (Result: exit-code) since Thu 2020-02-20 09:42:53 EST; 18s ago
  Process: 14314 ExecStart=/usr/local/calibre-web/cps.py -p /library/calibre-web/config/app.db (code=exited, status=1/FAILURE)
 Main PID: 14314 (code=exited, status=1/FAILURE)
>
> Feb 20 09:42:53 box.lan cps.py[14314]: INFO:cps.web:Performing restart of Calibre-Web
Feb 20 09:42:53 box.lan cps.py[14314]: Traceback (most recent call last):
Feb 20 09:42:53 box.lan cps.py[14314]:   File "/usr/local/calibre-web/cps.py", line 34, in <module>
Feb 20 09:42:53 box.lan cps.py[14314]:     from cps import create_app
Feb 20 09:42:53 box.lan cps.py[14314]:   File "/usr/local/calibre-web/cps/__init__.py", line 28, in <module>
Feb 20 09:42:53 box.lan cps.py[14314]:     from babel import Locale as LC
Feb 20 09:42:53 box.lan cps.py[14314]: ImportError: No module named babel
Feb 20 09:42:53 box.lan systemd[1]: calibre-web.service: Main process exited, code=exited, status=1/FAILURE
Feb 20 09:42:53 box.lan systemd[1]: calibre-web.service: Unit entered failed state.
Feb 20 09:42:53 box.lan systemd[1]: calibre-web.service: Failed with result 'exit-code'.

> root@box:~# systemctl start calibre-web

> root@box:~# systemctl status calibre-web
● calibre-web.service
   Loaded: loaded (/etc/systemd/system/calibre-web.service; enabled; vendor preset: enabled)
   Active: failed (Result: exit-code) since Thu 2020-02-20 09:44:14 EST; 1s ago
  Process: 18154 ExecStart=/usr/local/calibre-web/cps.py -p /library/calibre-web/config/app.db (code=exited, status=203/EXEC)
 Main PID: 18154 (code=exited, status=203/EXEC)
>
> Feb 20 09:44:14 box.lan systemd[1]: Started calibre-web.service.
Feb 20 09:44:14 box.lan systemd[1]: calibre-web.service: Main process exited, code=exited, status=203/EXEC
Feb 20 09:44:14 box.lan systemd[1]: calibre-web.service: Unit entered failed state.
Feb 20 09:44:14 box.lan systemd[1]: calibre-web.service: Failed with result 'exit-code'.

**Comment by holta** [comment]
Above `systemctl status calibre-web` output revised to be more comprehensive (earlier paste had cut off everything to the right of the 80th column!)

Ref: https://github.com/janeczku/calibre-web/issues/1206

**Comment by jvonau** [comment]
> Question 1:
> 
>     * Delete the root folder of calibre-web, everything is stored within this folder, unless you didn't change location of logfiles and settingsfile via command line. Depending on how you installed the dependencies you have to take care about them: delete the virtual environment folder, or the vendor folder (should be within calibre-web folder). If dependencies are installed systemwide via "pip" in the sites-packages folder you have to uninstall it (difficult to handle, as other programs also could need this).

The virtual environment is at /usr/local/calibre-web with the python dependencies in  /usr/local/calibre-web/lib/python-2.7|3.7/site-packages/, so it's just a matter of deleting /usr/local/calibre-web.

>       Don't forget to delete systemd scripts  and other files, you might created.

That would be just /etc/systemd/system/calibre-web.service.
>       Question 2:
>       The updater is quite stupid, just takes all the files from the new version (included in the zip file downloaded from the github page), replaces all files in the folder by the new files and deletes all files/folders not included anymore in the new zip file. EXCEPT: The settings file, the log files, the settings files for the gdrive integrtion and the vendor folder. I will add up the "venv" folder (I thought I already did this). I will try to improve the logic, so it will exclude folders within calibre.web where dependencies are loaded from.

From above think you just need to exclude lib/python* from the delete phase of the upgrade process.
>       For more ideas I need more details what exactly went wrong beside "...has not worked for me, in fact damaged the installation of Calibre-Web..."

I'm trying to help with the translation process.

**Comment by OzzieIsaacs** [comment]
It's done, release 0.6.6 of Calibre-Web is published.

**Comment by OzzieIsaacs** [comment]
I had a little problem with the release. The tag on github, where the release based on was the wrong commit. I corrected it now, so the release is now the right one. 
Sorry for the inconvenience.

**Comment by OzzieIsaacs** [comment]
I had a little problem with the release. The tag on github, where the release based on was the wrong commit. I corrected it now, so the release is now the right one. 
Sorry for the inconvenience.

**Comment by jvonau** [comment]
stopping service still hits timeout then killed
>  Active: deactivating (stop-sigterm) since Wed 2020-03-11 07:58:07 GMT; 2min 29s ago

> pi@box:/opt/iiab/iiab $ sudo systemctl status calibre-web.service --no-pager -l
● calibre-web.service - Calibre-Web
     Loaded: loaded (]8;;file://box.lan/etc/systemd/system/calibre-web.service/etc/systemd/system/calibre-web.service]8;;; enabled; vendor preset: enabled)
     Active: failed (Result: timeout) since Wed 2020-03-11 08:01:07 GMT; 1min 30s ago
    Process: 8004 ExecStart=/usr/local/calibre-web-py3/bin/python3 /usr/local/calibre-web-py3/cps.py -p /library/calibre-web/config/app.db (code=killed, signal=KILL)
   Main PID: 8004 (code=killed, signal=KILL)
Mar 09 17:32:40 box.lan systemd[1]: Started Calibre-Web.
Mar 11 07:58:07 box.lan systemd[1]: Stopping Calibre-Web...
Mar 11 08:01:07 box.lan systemd[1]: calibre-web.service: State 'stop-sigterm' timed out. Killing.
Mar 11 08:01:07 box.lan systemd[1]: calibre-web.service: Killing process 8004 (python3) with signal SIGKILL.
Mar 11 08:01:07 box.lan systemd[1]: calibre-web.service: Main process exited, code=killed, status=9/KILL
Mar 11 08:01:07 box.lan systemd[1]: calibre-web.service: Failed with result 'timeout'.
Mar 11 08:01:07 box.lan systemd[1]: Stopped Calibre-Web.

**Comment by OzzieIsaacs** [comment]
Doublechecked it, for me it's working on **raspberry pi zero**:
grafik
grafik
I'm also able to start and stop calibre-web on my normal **Linux Mint** PC with python 3.6.9. My **Ubuntu** server also with 3.6.9 is also starting and stopping calibre-web without any issue. (using version 0.6.6 upwards). I'm using the system environment without virtual environment but this shouldn't make any difference. My systemd unit file is the example given in the calibre-web wiki.

**Comment by jvonau** [comment]
The only difference I see is in the start line is the -p for the settings db, part of the original author's unit file, I didn't write it. The question becomes do we need to use -p in the start line @holta?

**Comment by OzzieIsaacs** [comment]
The -p only gives the location of the settingsfile. This shouldn‘t make a difference, on my Ubunutu server I‘m using it also, without any problem.

**Comment by OzzieIsaacs** [comment]
Which python3 version are you using?

**Comment by OzzieIsaacs** [comment]
The tornado version could also make a difference, I checked several versions from 4.x to the most recent 6.x version without problem. Are you really sure, you got the right version? The line that matters is in cps/server.py last line:
self.wsgiserver.add_callback_from_signal(self.wsgiserver.stop)
.._from_signal.. is the fix

**Comment by jvonau** [comment]
This test box has python 3.7.6 (raspbian-10 updated to unstable) that is why I was asking if Adam if this he saw this issue with raspbian-10 which uses python 3.7.3. Looks like the correct tornado version is insalled
> ls /usr/local/calibre-web-py3/lib/python3.7/site-packages/tornado*
/usr/local/calibre-web-py3/lib/python3.7/site-packages/tornado-6.0.3.dist-info

**Comment by jvonau** [comment]
It's python 3.7.6 causing the issue on my test box, just ran this install on a different machine and the service shuts down clean.
> iiab-admin@box:/opt/iiab/iiab $ sudo systemctl stop calibre-web.service 
iiab-admin@box:/opt/iiab/iiab $ sudo systemctl status calibre-web.service --no-pager -l
● calibre-web.service
   Loaded: loaded (]8;;file://box.lan/etc/systemd/system/calibre-web.service/etc/systemd/system/calibre-web.service]8;;; enabled; vendor preset: enabled)
   Active: inactive (dead) since Wed 2020-03-11 15:38:41 EDT; 8min ago
  Process: 14419 ExecStart=/usr/local/calibre-web-py3/bin/python3 /usr/local/calibre-web-py3/cps.py -p /library/calibre-web/config/app.db (code=exited, status=0/SUCCESS)
 Main PID: 14419 (code=exited, status=0/SUCCESS)
Mar 11 15:38:01 box.lan systemd[1]: Started calibre-web.service.
Mar 11 15:38:40 box.lan systemd[1]: Stopping calibre-web.service...
Mar 11 15:38:41 box.lan systemd[1]: calibre-web.service: Succeeded.
Mar 11 15:38:41 box.lan systemd[1]: Stopped calibre-web.service.

**Comment by OzzieIsaacs** [comment]
I checked it with 3.8.1 on Manjaro Linux, which also worked. I think your first installation has a problem

**Pull request merged**